src/lib/mina_base/parties_logic.ml

```diff
@@ -428,48 +420,75 @@ module Make (Inputs : Inputs_intf) = struct
   module Ps = Inputs.Parties
 
   let get_next_party
-      (current_stack : Ps.t) (* The stack for the most recent snapp *)
+      (current_forest : Ps.t) (* The stack for the most recent snapp *)
       (call_stack : Call_stack.t) (* The partially-completed parent stacks *) =
-    (* Invariant: [call_stack] only contains stacks. *)
-    let next_stack, next_call_stack =
-      let res = Call_stack.pop call_stack in
-      let next_stack =
-        Opt.or_default ~if_:Ps.if_ ~default:Ps.empty (Opt.map ~f:fst res)
-      in
-      let next_call_stack =
-        Opt.or_default ~if_:Call_stack.if_ ~default:Call_stack.empty
-          (Opt.map ~f:snd res)
-      in
-      (next_stack, next_call_stack)
-    in
     (* If the current stack is complete, 'return' to the previous
        partially-completed one.
     *)
-    let current_stack, call_stack =
-      let current_is_empty = Ps.is_empty current_stack in
-      ( Ps.if_ current_is_empty ~then_:next_stack ~else_:current_stack
+    let current_forest, call_stack =
+      let next_forest, next_call_stack =
+        (* Invariant: call_stack contains only non-empty forests. *)
+        let res = Call_stack.pop call_stack in
+        let next_forest =
+          Opt.or_default ~if_:Ps.if_ ~default:Ps.empty (Opt.map ~f:fst res)
+        in
+        let next_call_stack =
+          Opt.or_default ~if_:Call_stack.if_ ~default:Call_stack.empty
+            (Opt.map ~f:snd res)
+        in
+        (next_forest, next_call_stack)
+      in
+      (* TODO: I believe current should only be empty for the first party in
```

Comment:
Even then, I think we inject the first party into the `current_forest` rather than the `call_stack`. Perhaps this check is unneeded. I'm pro- deleting it and seeing if anything breaks :)